### PR TITLE
BackendFilter/Sort

### LINKED ISSUE #161: Backend FIlter/Sort
### Motivation

Currently we do all table sorting/filtering. While this works for now, it relies on the user's hardware specs and scales badly. We should move to server side sorting/filtering to address this.

### Deliverables

- Create a reusable utility in `core/` that applies specified pagination, sorting, and filter parameters to a given query
- Modify the get all routes for each relevant module to use this util to handle pagination, sorting, and filtering
  - These features should be opt in with sensible defaults.
- Create/modify tests for the get all routes
  - Extend/create http utils to consolidate logic, and use parameterized tests whenever possible

### Important Notes

- Try to be as type safe as possible, but if it gets too hacky plain dicts are fine

### PR DESCRIPTION
**Pull request by MatthewBarghout**
High-level motivation for making the changes
Added server-side pagination, sorting, and filtering to the /api/parties/ endpoint to improve performance and scalability as the database grows. Previously, all filtering and sorting happened client-side, which doesn't scale beyond a few hundred records.
Changes:
Core utilities (backend/src/core/query_utils.py):

Added reusable pagination, sorting, and filtering utilities for SQLAlchemy queries
Implemented PaginationParams, SortParam, FilterParam models with Pydantic validation
Created apply_query_params() to combine filters, sorting, and pagination with security whitelisting
Added get_total_count() to efficiently count filtered results before pagination

Party service (backend/src/modules/party/party_service.py):

Added get_parties_paginated() method with support for:

Optional pagination (page_number, page_size)
Sorting by party_datetime, location_id, contact_one_id, or id
Filtering by location_id and/or contact_one_id

Whitelisted allowed sort/filter fields for security

Party router (backend/src/modules/party/party_router.py):

Updated list_parties endpoint to accept query parameters for pagination, sorting, and filtering
All features are opt-in with sensible defaults (backward compatible)

Tests (backend/test/modules/party/test_party_list_features.py):

Added comprehensive test suite with 22 tests covering:

Pagination edge cases (empty DB, exact pages, remainders, beyond last page)
Sorting (ascending/descending by different fields)
Filtering (single and multiple filters)
Combined features (filter + sort + paginate together)

Bug fixes:

Fixed PartyTestUtils.next_dict() to avoid mutating the overrides dict (was causing test failures)
Fixed get_total_count() to properly count rows instead of columns

Closes #161

### CODE REVIEW

File: backend/test/modules/party/party_utils.py [hunk 1]
```diff
@@ -73,15 +73,18 @@ def generate_defaults(count: int) -> dict[str, Any]:
...
+            local_overrides["location_id"] = location.id
 
-        if "contact_one_id" not in overrides:
+        if "contact_one_id" not in local_overrides:
             student = await self.student_utils.create_one()
-            overrides["contact_one_id"] = student.account_id
+            local_overrides["contact_one_id"] = student.account_id
 
-        return await super().next_dict(**overrides)
+        return await super().next_dict(**local_overrides)
```
Comment (naasanov) [inline backend/test/modules/party/party_utils.py]: This is a good addition. In order to maintain consistency, make these changes in all functions that use overrides across all the util classes

File: backend/src/modules/party/party_service.py [hunk 2]
```diff
@@ -172,6 +178,111 @@ async def get_parties(self, skip: int = 0, limit: int | None = None) -> list[Par
         parties = result.scalars().all()
         return [party.to_dto() for party in parties]
 
+    async def get_parties_paginated(
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: We need filter/sort/pagination on all "get all" routes. If the logic is properly centralized within the utils function it should just require <=10 lines of configuration for each function

File: backend/src/modules/party/party_service.py [hunk 3]
```diff
@@ -172,6 +177,90 @@ async def get_parties(self, skip: int = 0, limit: int | None = None) -> list[Par
...
+        self,
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: We need filter/sort/pagination on all "get all" routes. If the logic is properly centralized within the utils function it should just require <=10 lines of configuration for each function

File: backend/test/utils/pagination_test_utils.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
+"""
+Reusable test utilities for pagination, sorting, and filtering endpoints.
+
+These utilities help test 'get all' routes across different modules without
+duplicating test logic.
+"""
```
Comment (naasanov) [inline backend/test/utils/pagination_test_utils.py]: I see that you made the utils but I don't see them being used anywhere

File: backend/src/modules/party/party_service.py [hunk 5]
```diff
@@ -172,6 +177,90 @@ async def get_parties(self, skip: int = 0, limit: int | None = None) -> list[Par
...
+        )
+
+        # Build filter params from dict
+        filter_params: list[FilterParam] = []
+        if filters:
+            for field, value in filters.items():
+                if value is not None:
+                    filter_params.append(
+                        FilterParam(field=field, operator=FilterOperator.EQUALS, value=value)
+                    )
```
Comment (naasanov) [inline backend/src/modules/party/party_service.py]: Now that we have an overarching filter function, we can make "get parties by (location/contact/date_range/etc.)" delegate to get_parties_paginated

**Review by naasanov — CHANGES_REQUESTED** [review]
There were some comments that were not addressed. 

The primary thing that I'm looking for is a util function that handles everything from parsing query params to applying filters/sorting/pagination to calculating metadata, returning the `PaginatedResponse[MyDto]`. Could even pass FastAPI's request object to avoid having a million parameters. I understand I'm not the one writing this up so if you think this is not possible or would be clunky, please lmk and tell me why you think so.

Beyond that, I'm looking for that util function to be applied to all get-all routes, and write tests for all of these routes using utils to avoid having bloated tests.

File: backend/src/modules/location/location_router.py [hunk 6]
```diff
@@ -88,17 +88,20 @@ async def get_place_details(
 
 @location_router.get("", response_model=PaginatedLocationResponse)
```
Comment (naasanov) [inline backend/src/modules/location/location_router.py]: We are also missing sort/filter for accounts table. This is for the admin dashboard and there is an accounts table as well, so we need sort/filter/pagination for it

File: backend/src/core/query_utils.py [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+        # Apply sort order
+        if sort_param.order == SortOrder.DESC:
+            query = query.order_by(desc(field))
+        else:
+            query = query.order_by(asc(field))
+
+    return query
+
+
+def apply_filters[ModelType: DeclarativeMeta](
```
Comment (naasanov) [inline backend/src/core/query_utils.py]: Not sure if this has functionality for date filtering. Correct me if im wrong
